### DISCUSSION TITLE
WebWorker communication

### DISCUSSION THREAD

**Discussion by M-Mueller** [emscripten-core/emscripten#16416] [Q&A] [answered]
I'm wondering how to pass the result of a detached thread back to the main thread. My naive solution looks like this:
```cpp
#include <thread>
#include <future>
#include <iostream>
#include <chrono>

#include <emscripten.h>
#include <emscripten/bind.h>

void compute_stuff() {
	std::thread thread([] {
		std::cout << "Thread ID: " << std::this_thread::get_id() << std::endl;
		std::this_thread::sleep_for(std::chrono::seconds(2));
		EM_ASM(postMessage("Done!")); // signal the main thread that we are done
	});
	thread.detach();
}

EMSCRIPTEN_BINDINGS(Loader) {
	emscripten::function("compute_stuff", &compute_stuff);
}
```
```js
var webworker = new Worker("ThreadTest.worker.js");
webworker.onmessage = (e) => {
  console.log("Received message: " + e.data);
};

ThreadTest().then((module) => {
  module.compute_stuff();
});
```
Compiled with
```
em++ -g -s -pthread -std=c++17 --bind -s MODULARIZE=1 -s EXPORT_NAME=ThreadTest -s ALLOW_MEMORY_GROWTH=1 -s PTHREAD_POOL_SIZE=1 -s ASSERTIONS=1 main.cpp -o ThreadTest.js
```
However, this doesn't work because emscripten already uses the message channel to control the WebWorker (it will print "worker sent an unknown command undefined"). The only solution I found so far, is using `MAIN_THREAD_EM_ASM` instead and call some global function, which I would like to avoid.

**Comment by sbc100** [comment, marked as answer]
I change is currently in flight that allows you to use the postMessage mechanism to communicate with the main thread: #16239.  Hopefully that will solve your issue?

**Reply by M-Mueller** [reply]
This looks exactly like what I was looking for, thank you! I will try this branch next week.

**Reply by M-Mueller** [reply]
The changes from the PR work for me.

I also realized that the code in my example wasn't actually doing what I thought it would do: I thought the `Worker`, I create in js, would be added to the thread pool, but this is not the case. I also found another workaround using some internals:

```js
ThreadTest().then((module) => {
  module.PThread.allocateUnusedWorker();
  const worker = module.PThread.unusedWorkers[0];
  module.PThread.loadWasmModuleToWorker(worker);

  var builtin_onmessage = worker.onmessage;
  worker.onmessage = (e) => {
    if (e.data.cmd == "custom") {
      console.log("Got custom message: ", e.data.message);
    } else {
      builtin_onmessage(e); // forward to original handler
    }
  };

  module.compute_stuff();
});
```
and
```cpp
EM_ASM(postMessage({ cmd: "custom", message: "Done!" }));
```
Compiled without `PTHREAD_POOL_SIZE`. However, this solution is probably brittle, since this is not documented as public API anywhere.
